json_encoder.go

```diff
@@ -168,6 +172,52 @@ func (enc *jsonEncoder) AddMarshaler(key string, obj LogMarshaler) error {
 	return err
 }
 
+func (enc *jsonEncoder) arrayBegin() {
```

Comment:
Can we call these `addArrayBegin`, etc.? To my eye, methods with noun names look like simple getters. Since these mutate the receiver, I'd prefer them to have verb names.